src/utility_stmt_hooking.c

```diff
@@ -689,20 +704,78 @@ PathmanCopyFrom(CopyState cstate, Relation parent_rel,
 }
 
 /*
- * COPY FROM does not support FDWs, emit ERROR.
+ * Init COPY FROM, if supported.
  */
 static void
-prepare_rri_for_copy(EState *estate,
-					 ResultRelInfoHolder *rri_holder,
-					 const ResultPartsStorage *rps_storage,
-					 void *arg)
+prepare_rri_for_copy(ResultRelInfoHolder *rri_holder,
+					 const ResultPartsStorage *rps_storage)
 {
-	ResultRelInfo  *rri = rri_holder->result_rel_info;
-	FdwRoutine	   *fdw_routine = rri->ri_FdwRoutine;
+	ResultRelInfo	*rri = rri_holder->result_rel_info;
+	FdwRoutine		*fdw_routine = rri->ri_FdwRoutine;
 
 	if (fdw_routine != NULL)
+	{
+		/*
+		 * If this Postgres has no idea about shardman, behave as usual:
+		 * vanilla Postgres doesn't support COPY FROM to foreign partitions.
+		 * However, shardman patches to core extend FDW API to allow it,
+		 * though currently postgres_fdw does so in a bit perverted way: we
+		 * redirect COPY FROM to parent table on foreign server, assuming it
+		 * exists, and let it direct tuple to proper partition. This is
+		 * because otherwise we have to modify logic of managing connections
+		 * in postgres_fdw and keep many connections open to one server from
+		 * one backend.
+		 */
+#ifndef PG_SHARDMAN
+		goto bail_out; /* to avoid 'unused label' warning */
+#else
```

Comment:
This is the most disputable change so far. My idea's that we could create a struct of function pointers and allocate it using `find_rendezvous_variable()`. This would allow us to move the FDW-related logic to pg_shardman. The only drawback of such approach is that pg_shardman would have to include the same definition of such struct. I see several options.